### ISSUE TITLE
Wiring and other ideas

### ISSUE THREAD

**Issue by knorke74** [samphonic/MoppyClassic#12]
Here are some ideas about the wiring:
- I measured that with my floppy drives ALL odd pins of the floppy connector BUT pin 1 are connected internally so if one of them is grounded with the Arduino everything works fine.
- The SELECT pin can be grounded by connecting it to one of the other odd pins. 
- This reduces the total number of wires between each floppy drive and the Arduino to 5 including power supply (STEP, DIRECTION, GND, 5V, GND)

Sammy, I'd really like to understand the java/midi part of your software, because I'd like to control some other acoustic devices like bells, little organ pipes or micro drum sets via midi and the Arduino, BUT I'm not very good at Java programming, so I can't understand what you did just by analyzing your code in NetBeans. Can you tell my something about your method or where you got the information on how to read and convert midi signals and how to send them to the Arduino? Thanks in advance, any help is much appreciated!

Regards,
Thorsten

**Comment by samphonic** [comment]
Thanks for the tips, I'll try to compile them into the Wiki on here if I get a chance (feel free to do so yourself... if... you can... I'm not sure how editing permissions work on github).

So like many things in Java (have you ever tried to do anything with JDBC?) the MIDI interfaces are a little bit non-intuitive.  The two big players are **MoppySequencer** and **MoppyPlayer**.

**MoppySequencer** implements a javax.sound.midi.Sequencer object that's used to actually play the MIDI files.  MIDI files are loaded using setSequence().  There's tutorials for setting this part up, and it's actually fairly straight-forward compared to the rest of the system.  The Sequencer object runs through the MIDI file when you call start() and traverses over each _event_ in the MIDI file.  By default, the Sequencer doesn't do much with these events, but you can utilize them by registering Listeners and Receivers.  **MoppySequencer** registers itself as a metaeventlistener to listen for events through the meta() method.  For Moppy, I use this to listen for tempo-change events and update the sequencer's tempo accordingly.

**MoppyPlayer** implements the MIDI Receiver interface, which means the send() method gets called for each MIDI note event.  Then I do a whole bunch of crazy things to figure out what the event was.  [This page](http://www.midi.org/techspecs/midimessages.php) was invaluable in figuring out what I needed.  Something to keep in mind here is that MIDI bytes are represented as 0-255, but Java stores bytes as -128-127, so there's some weird bitwise operations going on to get that settled.

For Moppy, I'm actually converting (in the MoppyPlayer.send() method) MIDI events into pin/period data (i.e. set pin2 to a loop-period of 4000, which is determined by the note-period divided by the Arduino's timer-interrupt delay).  The pin/period data is then sent over serial to the Arduino.

If you wanted to implement some custom code for handling events, I'd suggest you create your own class that implements the Receiver interface, and do your work in there.  You could actually even use the MoppyDesk application to run your code by switching out MoppyPlayer with your custom class on MoppySequencer:40 .

The **MoppyBridge** class just handles serializing the data and sending it over the serial port to the Arduino.  Right now messages as sent as a single byte and an integer, but you could customize that for your needs.

Let me know if you've got any other questions; I look forward to seeing what you end up making.

**Comment by jerossen** [comment]
Actually you only need 3 wires :) As pin 11 is also a ground pin, you can just connect your ground wire to pin 12 and 11. Like this http://georgewhiteside.net/projects/diskette-organ/dO-connector-wiring.jpg
(Got it from another floppy-organ project (http://georgewhiteside.net/projects/diskette-organ/))

**Comment by fructis** [comment]
Regarding wiring... Why do you need individual direction pins? Wouldn't a shared one be enough since you're pulsing the clock lines in sequence? Does the floppy controller need the direction pin to stay stable once clk has been pulsed?

About the ino code... Why aren't the 8 IFs in tick() enclosed in a for loop? Would be far more clear like this: for(i = 2; i < 17; i+=2) {... } and then just use i and i+1 for the addressing.

**Comment by fructis** [comment]
Furthermore, why the firstRun thing? Wouldn't it be simpler to put that bit of code in setup() ?

**Comment by samphonic** [comment]
> Regarding wiring... Why do you need individual direction pins? Wouldn't a shared one be enough since you're pulsing > the clock lines in sequence? Does the floppy controller need the direction pin to stay stable once clk has been 
> pulsed?

Individual pins are needed because all the drives are not always moving; and when they are, they are not necessarily moving at the same speed (i.e. some will reach the end of the track sooner than others) or in the same direction.

> About the ino code... Why aren't the 8 IFs in tick() enclosed in a for loop? Would be far more clear like this: for(i = 2; i > < 17; i+=2) {... } and then just use i and i+1 for the addressing.

It would indeed be more clear, but early on in the process I ran up against the limits of the Arduino's processing power, and the drives would start to stutter when more than two or three were playing at note at the same time.  I hard-coded each bit of the loop to eliminate any math that needed to be done (e.g. i+1, i+=2, &c.).  Ultimately, I switched to using an interrupt timer instead of the modulo function in the loop, and it might be possible to condense that into a for loop now.  I haven't gone back to change / test it only because it seems to work fine now and I didn't want to mess anyone up.

> Furthermore, why the firstRun thing? Wouldn't it be simpler to put that bit of code in setup() ?

I... don't know.  =P  This was my very first Arduino program, and I think I was still in the mindset that the setup() function was just for setting up pin states and loop() was where any sort of actions should go.  That firstRun stuff can almost certainly be eliminated; you're right.

**Comment by fructis** [comment]
Regarding the first question (hmm, can't find the quote button):
I don't see how that has anything to do with the direction pins. The steppers move due to the pulsing of the step line. If you don't want a stepper to move, don't pulse it's step line, regardless of the state of the direction pin. Again, the speed depends on the step line, not direction. I haven't really tried this project yet, so I'm assuming the floppy steppers behave this way, i.e. similar to L297 which I have experience with. So since you're not pulsing all the floppies' step lines at the same time, but rather one after the other, it wouldn't matter if the direction pin is common, since a floppy only looks at it's direction pin when it's clock (step) is pulsed. This wouldn't work if the floppy requires the direction pin to be stable after clock has been pulsed, but this is uncommon - stepper controllers usually contain a flip-flop. I'm suggesting all this since I have a bunch of ATtiny2313 laying around which could be adequate for the project if you condensed the required 16 lines into 9.

**Comment by samphonic** [comment]
(Yeah, I can't find a quote button either, I had to manually copy and paste...)

Ohhhhhhhh, I see what you're saying now.  Yeah, I think that would work just fine actually.  I'm probably not going to get around to making any code changes anytime soon (not to mention re-wiring my drives to share a direction pin), but if you or someone else does implement some of those changes, I'll definitely look to at least link to the code as an alternative, if not accept a pull request for them.

**Comment by hadrimoppy** [comment]
Hey we are building a Floppy / Boardunio Jukebox at the moment with the support of at least two Harddrives ;) It works really fine and sounds great! But is there a possibility to get some midi files from your project, so that we don't have to do double work in rebuilding them in midi? That would be very very great! 
And is there a possibility to make a playlist or a loop (I have absolutely no knowledge of java ;)). But thank you very very much for your documentation!! It helps a lot!
Best regards,
Harddrivemoppy ;)

**Comment by mamarley** [comment]
I too have build a floppy drive organ, and I have another idea for wiring.  The way I set mine up is that I used a computer power supply to power all the drives and I connected the ground from the power supply to the Arduino.  This means that I don't have to run a separate ground wire from each drive's data connector to the Arduino.  Each drive has four wires going to it, +5v and GND (from the power supply) and Step and Direction (from the Arduino).

**Comment by samphonic** [comment]
@hadrimoppy Send me an email at [redacted], let me know which MIDIs you'd like, and I'll see what I can do.  As for the playlist, YouTube user MrSolidSnake745 implemented some playlist functionality on his own, but he might be willing to share his code with you if you ask nicely.  I'm mostly occupied with other stuff, so I probably won't get around to making modifications anytime soon.

@mamarley That's an excellent tip!  Definitely would have saved me a lot of hassle trying to figure out grounding issues. =P

**Comment by solidsnake745** [comment]
@hadrimoppy This is really late and I think you've reached out to me already, but I made a new version of my playlist mod for Moppy Advanced. It's not complete though. I just did it in my spare time on a bunch of international flights >_> It's definitely way better than what I had before though. That one was pretty quickly put together.

Anyway, I forked off of Sammy's code if you'd like to check it out: https://github.com/solidsnake745/Moppy/tree/moppy-advanced
